django/utils/autoreload.py

```diff
@@ -198,6 +210,97 @@ def update_watch(sender=None, **kwargs):
     return EventHandler.modified_code
 
 
+def win32_notify_code_changed():
+    """
+    Check for changed code using Windows API. After being called
+    it blocks until a change event has been fired.
+    """
+    # WaitForMultipleObjects() can wait for only 64 handles max
+    # This is why cannot monitor individual files neither each parent
+    # directory as this would exceed the limit for a django project.
+    #
+    # For this reason watch directories, not files and also watch common
+    # parent directories to further limit the number of items.
+    common_parent_dirs = ['\\site-packages\\', '\\lib\\']
+
+    def reduce_path_to_common_dir(dirpath):
+        for dirname in common_parent_dirs:
+            if dirname in dirpath:
+                dirpath = dirpath[:dirpath.rfind(dirname) + len(dirname) - 1]
+                break
+        return dirpath
+
+    dirs = set()
+    watched_dirs = []
+    handles = []
+    for filename in gen_filenames():
+        dirpath = os.path.dirname(filename)
+        dirpath = dirpath if dirpath else '.'
+        dirpath = reduce_path_to_common_dir(dirpath)
```

Comment:
I’m not comfortable with how this is calculated. There ought to be a better way to handle this, e.g. by inspecting `sys.path` for common ancestors.